app/src/main/java/org/simple/clinic/user/registeruser/RegisterUser.kt

```diff
@@ -0,0 +1,92 @@
+package org.simple.clinic.user.registeruser
+
+import com.f2prateek.rx.preferences2.Preference
+import io.reactivex.Completable
+import io.reactivex.Single
+import org.simple.clinic.analytics.Analytics
+import org.simple.clinic.facility.Facility
+import org.simple.clinic.facility.FacilityRepository
+import org.simple.clinic.registration.RegistrationApi
+import org.simple.clinic.registration.RegistrationRequest
+import org.simple.clinic.registration.RegistrationResponse
+import org.simple.clinic.user.LoggedInUserPayload
+import org.simple.clinic.user.User
+import org.simple.clinic.user.User.LoggedInStatus.LOGGED_IN
+import org.simple.clinic.util.Just
+import org.simple.clinic.util.Optional
+import timber.log.Timber
+import java.io.IOException
+import java.util.UUID
+import javax.inject.Named
+
+class RegisterUser(
+    private val registrationApi: RegistrationApi,
+    private val userDao: User.RoomDao,
+    private val facilityRepository: FacilityRepository,
+    @Named("preference_access_token") private val accessTokenPreference: Preference<Optional<String>>
+) {
+
+  fun registerUserAtFacility(user: User, facility: Facility): Single<RegistrationResult> {
+    return Single.fromCallable { userToPayload(user, facility.uuid) }
+        .doOnSubscribe { Timber.i("Registering user") }
+        .map(::RegistrationRequest)
+        .flatMap(registrationApi::createUser)
+        .flatMap(::storeUserAndAccessToken)
+        .doOnSuccess(Analytics::setNewlyRegisteredUser)
+        .map { RegistrationResult.Success as RegistrationResult }
+        .onErrorReturn(::mapErrorToRegistrationResult)
+        .doOnSuccess { Timber.i("Registration result: $it") }
+
+  }
```

Comment:
WDYT?

```kotlin
  fun registerUserAtFacility(user: User, facility: Facility): Single<RegistrationResult> {
    val registrationRequest = RegistrationRequest(userToPayload(user, facility.uuid))
    return registrationApi
        .createUser(registrationRequest)
        .doOnSubscribe { Timber.i("Registering user") }
        .flatMap(::storeUserAndAccessToken)
        .doOnSuccess(Analytics::setNewlyRegisteredUser)
        .map { RegistrationResult.Success as RegistrationResult }
        .onErrorReturn(::mapErrorToRegistrationResult)
        .doOnSuccess { Timber.i("Registration result: $it") }
  }
```

Reply:
I just made sure,
1. All the side-effect operators `doOn*` are at the same level.
2. Got rid of `Single.fromCallable()` and start with the meat `registrationApi.createUser(registrationRequest)`.

Reply:
Looks good, I'll use this.

Reply:
I could use `.doOnError(Timber::e)` because that changes the behaviour. Earlier, we would not report `IOException` errors since those are expected, but this would end up reporting those as well.

I added another method to report errors which would filter them out before reporting.